src/common/components/signUpModal/signUpModal.component.spec.js

```diff
@@ -2,132 +2,268 @@ import angular from 'angular'
-          })
-        })
+  describe('loadDonorDetails()', () => {
+    it('should set loadingDonorDetails to false', () => {
+      jest.spyOn($ctrl.orderService, 'getDonorDetails').mockReturnValue(Observable.throw({status: 404}))
+      $ctrl.loadingDonorDetails = true
+      $ctrl.loadDonorDetails()
+      expect($ctrl.loadingDonorDetails).toEqual(false)
+    })
+  });
 
-        describe('500 Bad Request', () => {
-          it('sets generic error', () => {
-            deferred.reject({ status: 500 })
-            $rootScope.$digest()
+  describe('submitDetails()', () => {
+    it('should return as form is not valid', () => {
+      $ctrl.submitDetails()
+      expect($ctrl.signUpForm.$setSubmitted).toHaveBeenCalled()
+      expect($ctrl.submitting).toEqual(false)   
+    })
 
-            expect($ctrl.hasError).toEqual(true)
-            expect($ctrl.signUpErrors).toEqual(expect.objectContaining({ generic: true }))
-          })
-        })
+    it('should not call createAccount', async () => {
+      jest.spyOn($ctrl.sessionService, 'createAccount')
+
+      jest.spyOn($ctrl.orderService, 'getDonorDetails').mockImplementation(() => Observable.from(
+        [signUpFormData]
+      ))
+      $ctrl.$onInit()
+      $ctrl.signUpForm.$valid = false;
+      $ctrl.submitDetails(signUpFormData.email, signUpFormData.name['given-name'], signUpFormData.name['family-name']).then(() =>{
+        expect($ctrl.sessionService.createAccount).not.toHaveBeenCalled()
+      })
+    });
+
+    it('should redirect as already logged in', async () => {
+      jest.useFakeTimers();
+      jest.spyOn(global, 'setTimeout');
+      jest.spyOn($ctrl.sessionService, 'createAccount').mockImplementation(() => Promise.resolve({
+        status: 'error',
+        redirectToSignIn: true,
+        data: ['Error one', 'Error 2']
+      }))
+      
+
+      jest.spyOn($ctrl.orderService, 'getDonorDetails').mockImplementation(() => Observable.from(
+        [signUpFormData]
+      ))
+      $ctrl.$onInit()
+      $ctrl.signUpForm.$valid = true
+      $ctrl.submitDetails().then(() =>{
+        expect($ctrl.$scope.$apply).toHaveBeenCalled();
+        expect($ctrl.sessionService.createAccount).toHaveBeenCalledWith(signUpFormData.email, signUpFormData.name['given-name'], signUpFormData.name['family-name'])
+        expect($ctrl.submitting).toEqual(false)
+        expect($ctrl.submissionError).toEqual(['Error one', 'Error 2']);
+        jest.advanceTimersByTime(10000);
+        expect($ctrl.onStateChange).toHaveBeenCalledWith({ state: 'sign-in' })
+        jest.useRealTimers()
+      });
+    })
+
+    it("should call createAccount() with the user's data", () => {
+      jest.spyOn($ctrl.orderService, 'getDonorDetails').mockImplementation(() => Observable.from(
+        [signUpFormData]
+      ))
+      jest.spyOn($ctrl.sessionService, 'createAccount').mockImplementation(() => Promise.resolve({
+        status: 'error',
+        data: ['Error one', 'Error 2']
+      }))
```

Comment:
Would this not be a `Promise.reject`?

Reply:
This is because `sessionService.createAccount` returns this function`sessionService.catchCreateAccountErrors` when it handles errors, which returns a resolved promise.